### ISSUE TITLE
Error when running with gevent in Flask application

### ISSUE THREAD

**Issue by Shir0kamii** [pallets/werkzeug#830]
Hi,

It seems Werkzeug in its latest version can't work with gevent.

I have a simple Flask application and I have an exception just from lauching it : 

```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib64/python3.4/threading.py", line 920, in _bootstrap_inner
    self.run()
  File ".venv/lib/python3.4/site-packages/gevent/threading.py", line 137, in run
    super(Thread, self).run()
  File "/usr/lib64/python3.4/threading.py", line 868, in run
    self._target(*self._args, **self._kwargs)
  File ".venv/lib/python3.4/site-packages/werkzeug/serving.py", line 653, in inner
    fd=fd).serve_forever()
  File ".venv/lib/python3.4/site-packages/werkzeug/serving.py", line 557, in make_server
    passthrough_errors, ssl_context, fd=fd)
  File ".venv/lib/python3.4/site-packages/werkzeug/serving.py", line 465, in __init__
    socket.SOCK_STREAM)
  File ".venv/lib/python3.4/site-packages/gevent/_socket3.py", line 419, in fromfd
    nfd = dup(fd)
OSError: [Errno 9] Bad file descriptor
```

This error happens in version 0.11.2 but not in version 0.11.1. I use gevent in version 1.1b6 and python in version 3.4.3.

It doesn't happen if I get rid of the gevent's monkey patching and it doesn't occur during it, but rather during the execution of the `run` method of the flask application.

Does anyone can explain what is happening ?

**Comment by ramnes** [comment]
:+1:

**Comment by mitsuhiko** [comment]
It would appear gevent does not support `socket.fromfd`. Does this happen on Python 3 only or on 2.7 as well?

**Comment by ramnes** [comment]
Hey @mitsuhiko, thanks for the fast answer.

I have a slightly different traceback, but which happen in the same conditions, with `Werkzeug==0.11.3` or `Werkzeug==0.11.2`, `gevent==1.1b6` and a `gevent.monkey.patch_all()` on a Flask application:

```
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
 * Restarting with stat
 * Debugger is active!
 * Debugger pin code: 141-649-665
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib64/python3.4/threading.py", line 920, in _bootstrap_inner
    self.run()
  File ".venv/lib/python3.4/site-packages/gevent/threading.py", line 137, in run
    super(Thread, self).run()
  File "/usr/lib64/python3.4/threading.py", line 868, in run
    self._target(*self._args, **self._kwargs)
  File ".venv/lib/python3.4/site-packages/werkzeug/serving.py", line 654, in inner
    fd=fd)
  File ".venv/lib/python3.4/site-packages/werkzeug/serving.py", line 548, in make_server
    passthrough_errors, ssl_context, fd=fd)
  File ".venv/lib/python3.4/site-packages/werkzeug/serving.py", line 473, in __init__
    self.server_address = self.socket.getsockname()
OSError: [Errno 88] Socket operation on non-socket
```

In my case, it only happens in debug mode, when the application is getting reloaded, and only with Python 3 (only tried 3.4 for now).

You can find my test case here: https://github.com/ramnes/werkzeug-830

**Comment by Shir0kamii** [comment]
Hi,

The test case work for me, even though I still have the same error `Bad file descriptor`.

Also it only happens on Python 3.

**Comment by ramnes** [comment]
The problem seems fixed with `gevent==1.2.1`, `Werkzeug==0.12.1` and `Flask==0.12.1`.
